Mathlib/NumberTheory/LSeries/ZMod.lean

```diff
@@ -152,49 +183,382 @@ lemma LFunction_stdAddChar_eq_expZeta (j : ZMod N) (s : ℂ) (hjs : j ≠ 0 ∨
     neg_sub]
   congr 2
   -- now gather sum terms
-  rw [LFunction_dft _ hs', LFunction_dft _ hs']
+  rw [LFunction_dft _ hs', LFunction_dft _ (hs'.imp_left <| by simp only [neg_zero, imp_self])]
   conv_rhs => enter [2, 2]; rw [← (Equiv.neg _).sum_comp _ _ (by simp), Equiv.neg_apply]
   simp_rw [neg_neg, mul_sum, ← sum_add_distrib, ← mul_assoc, mul_comm _ (Φ _), mul_assoc,
     ← mul_add, map_neg, add_comm]
 
+section signed
+
+variable {Φ : ZMod N → ℂ}
+
+lemma LFunction_def_even (hΦ : Φ.Even) (s : ℂ) :
+    LFunction Φ s = N ^ (-s) * ∑ j : ZMod N, Φ j * hurwitzZetaEven (toAddCircle j) s := by
+  simp only [LFunction, hurwitzZeta, mul_add (Φ _), sum_add_distrib]
+  congr 1
+  simp only [add_right_eq_self, ← neg_eq_self ℂ, ← sum_neg_distrib]
+  refine Fintype.sum_equiv (.neg _) _ _ fun i ↦ ?_
+  simp only [Equiv.neg_apply, hΦ i, map_neg, hurwitzZetaOdd_neg, mul_neg]
+
+lemma LFunction_def_odd (hΦ : Φ.Odd) (s : ℂ) :
+    LFunction Φ s = N ^ (-s) * ∑ j : ZMod N, Φ j * hurwitzZetaOdd (toAddCircle j) s := by
+  simp only [LFunction, hurwitzZeta, mul_add (Φ _), sum_add_distrib]
+  congr 1
+  simp only [add_left_eq_self, ← neg_eq_self ℂ, ← sum_neg_distrib]
+  refine Fintype.sum_equiv (.neg _) _ _ fun i ↦ ?_
+  simp only [Equiv.neg_apply, hΦ i, map_neg, hurwitzZetaEven_neg, neg_mul]
+
+/-- Explicit formula for `LFunction Φ 0` when `Φ` is even. -/
+lemma LFunction_apply_zero_of_even (hΦ : Φ.Even) :
+    LFunction Φ 0 = -Φ 0 / 2 := by
+  simp only [LFunction_def_even hΦ, neg_zero, cpow_zero, hurwitzZetaEven_apply_zero,
+    toAddCircle_eq_zero, mul_ite, mul_div, mul_neg_one, mul_zero, sum_ite_eq', Finset.mem_univ,
+    ↓reduceIte, one_mul]
+
+/-- The L-function of an even function vanishes at negative even integers. -/
+lemma LFunction_neg_two_mul_nat_add_one (hΦ : Φ.Even) (n : ℕ) :
+    LFunction Φ (-2 * (n + 1)) = 0 := by
+  simp only [LFunction_def_even hΦ, hurwitzZetaEven_neg_two_mul_nat_add_one, mul_zero,
+    sum_const_zero]
+
+/-- The L-function of an odd function vanishes at negative odd integers. -/
+lemma LFunction_neg_two_mul_nat_sub_one (hΦ : Φ.Odd) (n : ℕ) :
+    LFunction Φ (-2 * n - 1) = 0 := by
+  simp only [LFunction_def_odd hΦ, hurwitzZetaOdd_neg_two_mul_nat_sub_one, mul_zero, sum_const_zero]
+
+variable (Φ) in
+/--
+The completed L-function of a function mod `N`.
+
+This is only mathematically meaningful if `Φ` is either even, or odd. We extend this to all `Φ`
+as follows: if `Φ` is not even, then the completed L-series of `Φ` is the `L`-series of the odd
+part of `Φ`.
+-/
+noncomputable def completedLFunction (s : ℂ) : ℂ :=
+  if Φ.Even then N ^ (-s) * ∑ j, Φ j * completedHurwitzZetaEven (toAddCircle j) s
+  else N ^ (-s) * ∑ j, Φ j * completedHurwitzZetaOdd (toAddCircle j) s
+
+lemma completedLFunction_zero (s : ℂ) : completedLFunction (0 : ZMod N → ℂ) s = 0 := by
+  simp only [completedLFunction, Pi.zero_apply, zero_mul, sum_const_zero, mul_zero, ite_self]
+
+lemma completedLFunction_def_even (hΦ : Φ.Even) (s : ℂ) :
```

Comment:
Maybe this and various other lemms can be called `Even.blahbla` to have dot notation.

Reply:
I'm not so keen on using dot-notation for these, because I thought it was better to keep everything together in the `ZMod` namespace. These results make sense only for functions `ZMod N → ℂ` while `Function.Even` is much more general.

Reply:
OK, no problem.